### PR TITLE
Use click to parse arguments

### PR DESCRIPTION
**Pull request by 0xArty**
### Background
This PR changes the argument parsing library from `argparse` to `click`. The primary reason for this switch is to leverage the benefits provided by `click`, while maintaining backward compatibility with the existing implementation. `Click` is a powerful and flexible library for creating command-line interfaces, offering several advantages over `argparse`, such as a more intuitive and consistent API, better error messages, automatic generation of help messages, and advanced features like command chaining and nested commands.

### Changes
- Replace **argparse** with **click** for argument parsing.
- Update the **main** function to use **click** decorators and options.
- Renamed the **parse_arguments** to **create_config**
- Ensure that existing functionality is preserved with minimal changes to the codebase.
- Keep the door open for future enhancements using **click's** built-in features.

### Benefits of Click
- More intuitive and consistent API.
- Better error messages and automatic generation of help messages.
- Advanced features like command chaining and nested commands.
- Actively maintained and widely used library, with extensive documentation and community support.

### Documentation
The code changes include in-code comments to describe the new **click** decorators and options. The external documentation will be updated to reflect the change from **argparse** to **click**.

### Test Plan
1. Manually test the script with various command-line options to ensure that the existing functionality is preserved.
2. Run existing test cases to ensure that the change does not introduce any regressions.

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

**Comment by 0xArty** [comment]
Please see my previous attempt to get this in https://github.com/Significant-Gravitas/Auto-GPT/pull/1184

**Review by Pwuts — DISMISSED** [review]
Nice improvement to the CLI!

One small request: since there is only one command as of now, it's best if people can still run `./run.sh` without any subcommands. Click seems to support this with [click-default-group](https://github.com/click-contrib/click-default-group).

**Comment by 0xArty** [comment]
@Pwuts It does support running without subcommands. However, one of the key reasons I feel we should use click is the ability to add sub-commands. 

Once this is merged I expect to start adding sub-commands fairly quickly. Such as:
```
./run.sh pulgins  list
./run.sh pulgins activate / deactivate
./run.sh research async-agent
```
etc

**Comment by 0xArty** [comment]
> Nice improvement to the CLI!
> 
> One small request: since there is only one command as of now, it's best if people can still run `./run.sh` without any subcommands. Click seems to support this with [click-default-group](https://github.com/click-contrib/click-default-group).

I'm not sure it's wise using a dependence with only 50 stars and hasn't been touched for 4 years

**Comment by Pwuts** [comment]
It has good test coverage, and we could pin it to the current version if we want to be certain it keeps working. There are also ways to achieve the same in bash, but the lib seems like a cleaner option.

**Comment by 0xArty** [comment]
@Pwuts Figured out how to do it without adding another dependency:

```bash
➜  Auto-GPT git:(click-arg-pasing) ✗ ./run.sh 
All packages are installed.
Welcome to Auto-GPT!  run with '--help' for more information.
Create an AI-Assistant:  Enter the name of your AI and its role below. Entering nothing will load defaults.
Name your AI:  For example, 'Entrepreneur-GPT'
AI Name: 
```
help:

```
➜  Auto-GPT git:(click-arg-pasing) ✗ ./run.sh --help
All packages are installed.
Usage: python -m autogpt [OPTIONS] COMMAND [ARGS]...

  Welcome to AutoGPT an experimental open-source application showcasing the
  capabilities of the GPT-4 pushing the boundaries of AI.

  Start an Auto-GPT assistant.

Options:
  -c, --continuous                Enable Continuous Mode
  -y, --skip-reprompt             Skips the re-prompting messages at the
                                  beginning of the script
  -C, --ai-settings TEXT          Specifies which ai_settings.yaml file to
                                  use, will also automatically skip the re-
                                  prompt.
  -l, --continuous-limit INTEGER  Defines the number of times to run in
                                  continuous mode
  --speak                         Enable Speak Mode
  --debug                         Enable Debug Mode
  --gpt3only                      Enable GPT3.5 Only Mode
  --gpt4only                      Enable GPT4 Only Mode
  -m, --use-memory TEXT           Defines which Memory backend to use
  -b, --browser-name TEXT         Specifies which web-browser to use when
                                  using selenium to scrape the web.
  --allow-downloads               Dangerous: Allows Auto-GPT to download files
                                  natively.
  --help                          Show this message and exit.

Commands:
  test  An example sub command
```
Here is some sample code showing how a subcommand can be added:

```python
@main.command()
def test() -> None:
    """An example sub command"""
    print("test")
```
Running with test passed as the subcommand:
```
➜  Auto-GPT git:(click-arg-pasing) ✗ ./run.sh test --help
All packages are installed.
Usage: python -m autogpt test [OPTIONS]

  An example sub command

Options:
  --help  Show this message and exit.
 ```

**Comment by jayceslesar** [comment]
click is also good because it handles some cases that argparse implied such as list arguments which is the biggest PITA with argparse (there is some nasty implied ordering in argparse)... 

If you wanted to add some tests could do something like
```python
import click
from click.testing import CliRunner

def test_cli():
    runner = CliRunner()
    result = runner.invoke(path.to.cli, ["my-cli-command", "my-click-options", "my-click-args"])
    assert result.exit_code == 0
```

**Review by merwanehamadi — APPROVED** [review]

**Review by p-i- — APPROVED** [review]

File: autogpt/__main__.py [hunk 1]
```diff
@@ -73,51 +67,56 @@ def start(
...
-        ai_name=ai_name,
-        memory=memory,
-        full_message_history=full_message_history,
-        next_action_count=next_action_count,
-        system_prompt=system_prompt,
-        triggering_prompt=triggering_prompt,
-    )
-    agent.start_interaction_loop()
+    """
+    Welcome to AutoGPT an experimental open-source application showcasing the capabilities of the GPT-4 pushing the boundaries of AI.
```
Comment (Pwuts) [inline autogpt/__main__.py]: punctuation?

**Pull request merged**